server/src/main/java/org/elasticsearch/tasks/Task.java

```diff
@@ -41,17 +41,14 @@
      */
     public static final String X_ELASTIC_PRODUCT_ORIGIN_HTTP_HEADER = "X-elastic-product-origin";
 
-    public static final Set<String> HEADERS_TO_COPY = Set.of(
-        X_OPAQUE_ID_HTTP_HEADER,
-        TRACE_PARENT_HTTP_HEADER,
-        X_ELASTIC_PRODUCT_ORIGIN_HTTP_HEADER
-    );
     /**
      * Parsed part of traceparent. It is stored in thread context and emitted in logs.
      * Has to be declared as a header copied over for tasks.
      */
     public static final String TRACE_ID = "trace.id";
 
+    public static final Set<String> HEADERS_TO_COPY = Set.of(X_OPAQUE_ID_HTTP_HEADER, TRACE_ID, X_ELASTIC_PRODUCT_ORIGIN_HTTP_HEADER);
```

Comment:
The confusion and mistake was because for thread context we want to copy x-opaque-id, x-elastic-product-origin and trace.id https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/node/Node.java#L729

however for headers from RestRequest we want to copy x-opaque-id, x-elastic-product-origin and traceparent
https://github.com/elastic/elasticsearch/blob/a2bc4854b562782a8a66eeacb23246d4d21a2b01/server/src/main/java/org/elasticsearch/action/ActionModule.java#L460

traceparent is then parsed and sets trace.id on thread context.